diff --git a/toolchains/lowrisc_rv32imcb/repository.bzl b/toolchains/lowrisc_rv32imcb/repository.bzl
index 29acf96..8cea6ed 100644
--- a/toolchains/lowrisc_rv32imcb/repository.bzl
+++ b/toolchains/lowrisc_rv32imcb/repository.bzl
@@ -8,8 +8,8 @@ def lowrisc_rv32imcb_repos(local = None):
     http_archive_or_local(
         name = "lowrisc_rv32imcb_files",
         local = local,
-        url = "https://github.com/lowRISC/lowrisc-toolchains/releases/download/20240923-1/lowrisc-toolchain-rv32imcb-20240923-1.tar.xz",
-        sha256 = "aeea1983553f4c81c6409abcf0d6ca33b5ed4716b2b694e7ff030523cf13486a",
-        strip_prefix = "lowrisc-toolchain-rv32imcb-20240923-1",
+        url = "https://github.com/lowRISC/lowrisc-toolchains/releases/download/20250710-1/lowrisc-toolchain-rv32imcb-x86_64-20250710-1.tar.xz",
+        sha256 = "6f02aae27c097c71a2875a215896a0301e32ab56d6d26e917dae59d124c573fb",
+        strip_prefix = "lowrisc-toolchain-rv32imcb-x86_64-20250710-1",
         build_file = Label("//toolchains:BUILD.export_all.bazel"),
     )
